csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -320,10 +341,20 @@ public void Add(IEnumerable<T> values)
             {
                 throw new ArgumentNullException("values");
             }
-            // TODO: Check for ICollection and get the Count, to optimize?
-            foreach (T item in values)
+
+            ICollection collection = values as ICollection;
+            if (collection != null)
             {
-                Add(item);
+                this.EnsureSize(this.count + collection.Count);
+                collection.CopyTo(this.array, this.count);
```

Comment:
I don't think this is valid, unless T is a value type - because the collection could have null entries which should cause an exception to be thrown. (Even with value types, there are some potential oddities where an `int[]` counts as an `IEnumerable<uint>` at the CLR level, which then causes problems when copying to arrays.)